pcsx2/gui/Dialogs/GSDumpDialog.cpp

```diff
@@ -0,0 +1,798 @@
+
+	Bind(wxEVT_LIST_ITEM_SELECTED, &Dialogs::GSDumpDialog::SelectedDump, this, ID_DUMP_LIST);
+	Bind(wxEVT_BUTTON, &Dialogs::GSDumpDialog::RunDump, this, ID_RUN_DUMP);
+	Bind(wxEVT_BUTTON, &Dialogs::GSDumpDialog::ToStart, this, ID_RUN_START);
+	Bind(wxEVT_BUTTON, &Dialogs::GSDumpDialog::StepPacket, this, ID_RUN_STEP);
+	Bind(wxEVT_BUTTON, &Dialogs::GSDumpDialog::ToCursor, this, ID_RUN_CURSOR);
+	Bind(wxEVT_BUTTON, &Dialogs::GSDumpDialog::ToVSync, this, ID_RUN_VSYNC);
+	Bind(wxEVT_TREE_SEL_CHANGED, &Dialogs::GSDumpDialog::ParsePacket, this, ID_SEL_PACKET);
+	Bind(wxEVT_CHECKBOX, &Dialogs::GSDumpDialog::CheckDebug, this, ID_DEBUG_MODE);
+}
+
+void Dialogs::GSDumpDialog::GetDumpsList()
+{
+	wxDir snaps(g_Conf->Folders.Snapshots.ToAscii());
+	wxString filename;
+	bool cont = snaps.GetFirst(&filename, "*.gs", wxDIR_DEFAULT);
+	int i = 0, h = 0, j = 0;
+	m_dump_list->AppendColumn("Dumps");
+	// set the column size to be exactly of the size of our list
+	m_dump_list->GetSize(&h, &j);
+	m_dump_list->SetColumnWidth(0, h);
+
+	while (cont)
+	{
+		m_dump_list->InsertItem(i, filename.substr(0, filename.find_last_of(".")));
+		i++;
+		cont = snaps.GetNext(&filename);
+	}
+}
+
+void Dialogs::GSDumpDialog::SelectedDump(wxListEvent& evt)
+{
+	wxString filename_preview = g_Conf->Folders.Snapshots.ToAscii() + ("/" + evt.GetText()) + ".png";
+	wxString filename = g_Conf->Folders.Snapshots.ToAscii() + ("/" + evt.GetText()) + ".gs";
+	if (wxFileExists(filename_preview))
+	{
+		auto img = wxImage(filename_preview);
+		img.Rescale(250 * MSW_GetDPIScale(), 200 * MSW_GetDPIScale(), wxIMAGE_QUALITY_HIGH);
+		m_preview_image->SetBitmap(wxBitmap(img));
+		delete m_selected_dump;
+		m_selected_dump = new wxString(filename);
+	}
+	else
+	{
+		m_preview_image->SetBitmap(EmbeddedImage<res_NoIcon>().Get());
+	}
+}
+
+void Dialogs::GSDumpDialog::RunDump(wxCommandEvent& event)
+{
+	m_debug_mode->Enable();
+	m_start->Enable();
+	m_step->Enable();
+	m_selection->Enable();
+	m_vsync->Enable();
+	GetCorePlugins().Shutdown();
+	m_thread->Start();
+	return;
+}
+
+void Dialogs::GSDumpDialog::ProcessDumpEvent(GSData event, char* regs)
+{
+    switch (event.id)
+	{
+		case Transfer:
+		{
+			switch (event.path)
+			{
+				case Path1Old:
+				{
+					u32* data = (u32*)malloc(16384);
+					int addr = 16384 - event.length;
+					memcpy(data, event.data + addr, event.length);
+					GSgifTransfer1(data, addr);
+					free(data);
+					break;
+				}
+				case Path1New:
+				{
+					GSgifTransfer((u32*)event.data, event.length / 16);
+					break;
+				}
+				case Path2:
+				{
+					GSgifTransfer2((u32*)event.data, event.length / 16);
+					break;
+				}
+				case Path3:
+				{
+					GSgifTransfer3((u32*)event.data, event.length / 16);
+					break;
+				}
+				default:
+					break;
+			}
+			break;
+		}
+		case VSync:
+		{
+			GSvsync((*((int*)(regs + 4096)) & 0x2000) > 0 ? (u8)1 : (u8)0);
+			break;
+		}
+		case ReadFIFO2:
+		{
+			u64* arr = (u64*)malloc(*((int*)event.data));
```

Comment:
I went to the shared_ptr route instead, since the pointer would get out of scope when assigning it to my struct, even after an std::move since I was calling a function and it was starting to look like a mess.

std::vectors, as you said so well, 0 initialize and have a bigger overhead; knowing I will never resize my arrays I don't need all the semantics it provides.

On to the alloc of elements during the loop: can't get around it, I can't predict the size all elements will take and trying to just alloc the biggest block possible seems like a huge hack for me.